### ISSUE TITLE
Plea (high priority for me): interpret a "long click" as alt+left click

### ISSUE THREAD

**Issue by david-a-wheeler** [expln/metamath-lamp#54]
The metamath-lamp UI includes many uses of alt+left click. This action is *not* possible on typical smartphones to my knowledge. As far as I can tell this is the *only* UI mechanism that makes it not possible to fully use metamath-lamp on smartphones. In particular, you currently can't change a provable statement into a hypothesis if you can't use alt+left click. (Yes, you could create a JSON file and load it in, but that's a terrible workaround.) You also can't use the statement selector on smartphones, which is unfortunate because it's the device where keyboard use is the most annoying. Finally, sometimes alt+left click is inconvenient even when you're using a laptop/desktop.

I would *really* like to see a "long click" interpreted as alt+left click. That is, it would be considered alt+left click if the user presses the touchscreen or left button down, holds it down for a long time (say at least 1500 milliseconds) on the same element, and then it's released without moving off the selected element. That one addition would be enough to use everything on a smartphone, *and* isn't hard to understand.

I earlier noted this in #28, but after thinking about it, I've decided that I would *really* like this feature.

As far as I know React doesn't *directly* support longpress, but it doesn't appear to be hard to implement. Here's a discussion and sample code:
https://stackoverflow.com/questions/48048957/react-long-press-event

The "time for long press" could be a configurable setting in Settings. I'm guessing 1500 milliseconds would be a good default.

Note: this means a left-click on a not-being-edited proof statement would edit, but a long-click on a not-being-edited proof statement would invoke the statement selector. I don't think that will cause any problems, because both left-click and long-click only happen on touch/mouse press *up*. In short, the system can easily tell the difference; it's an alt-click if if the mouse button was held down on the element for at least "time for long press" time.

Ideally there'd be a visual indicator when the press-down-time had been exceeded that could be seen by a smartphone user. So I don't think it can be just a cursor image change (which would be nice but the user's digit would hide that). It could be a screen flash (e.g., briefly set the background to another color) or some other simple visual indicator; just pick something easy and distinct. That said, adding the visual indicator could be done later. My concern currently is to just have something that *works*.

This is a relatively high priority for me, because it suddenly means everything works on a smartphone. That doesn't make it necessarily a high priority for *you*, but I figure I can beg :-).

**Comment by expln** [comment]
> you could create a JSON file and load it in, but that's a terrible workaround

I even never thought about this :))

For me it will be much easier (and faster) to implement a so called "mobile mode" which may be turned on/off similar to "visualization on/off". In mobile mode clicking (touching) a statement will open a pop up with three available options to continue: change statement type, edit statement, launch context aware selection. You will have to chose required action and this action will be launched. Would this be acceptable?

**Comment by david-a-wheeler** [comment]
A "mobile mode" would work, but that sounds like a terrible UI.

I did find a tutorial specifically on how to implement long presses in React. It looks easier to implement than a "mobile mode", and probably more user-friendly:

https://spacejelly.dev/posts/how-to-detect-long-press-gestures-in-javascript-events-in-react/

**Comment by expln** [comment]
Ok, I will start implementing long touch and I will see how it goes. Testing should not be a problem, because Chrome, for example, can switch to mobile mode - open dev tools (F12) and you'll see a button in shape of "a smartphone in front of a tablet". When you click it, Chrome shows you the current site as if it were seen on a mobile device. You can even select a type of the mobile device.

**Comment by david-a-wheeler** [comment]
Thanks so much!

It's helpful to have UIs be as similar as possible from the POV of trying to explain them. I can't help that some keyboards use Opt while others use Alt, but if they otherwise work identically it's no big deal. But when dialogues show up on some systems and not others that is more complicated.

You're right, you could use a browser mode to test on mobile. But the more everything works the same way everywhere, the fewer surprises for users.

**Comment by expln** [comment]
@david-a-wheeler 

Hi David,

I deployed the long-click feature to dev https://expln.github.io/lamp/dev/index.html Please check how it works for you. Few notes about the change:

This change applies to: statements in the editor, statement types (H/P), the description.

long-click (long-touch) is now interpreted as alt-click - you may use both of them with same effect (long-touch on a touchscreen and long-click with a mouse).

The code for long-click is a bit complex, therefore it may contain difficult to spot bugs. So I made this feature optional. It is On by default but may be turned off if it causes any critical issue for a user.

The default long-click delay is 500 ms. I tested with it, it works well for me. Please let me know if you think it should be increased.

The long-click delay value is bounded below by 400ms and above by 3000ms. The lower bound is technical, I need it to ensure correct processing of some use cases. I can make it even less, but I don't think this is needed. The upper bound is "artificial", I introduced it to eliminate cases when a user, for example, puts an excessive 0 (puts 10000 instead of 1000) and then doesn't understand why it doesn't work. When a user puts an incorrect value of long-click delay then mm-lamp silently replaces it with a correct value when settings are saved.

I noticed that pressing a statement type indicator (H/P) is not convenient on a smartphone because it is too small and densely surrounded by other UI elements. I will try to fix this when I will work on space usage improvement.

>Ideally there'd be a visual indicator when the press-down-time had been exceeded that could be seen by a smartphone user.

I didn't implement any indicator, you just keep pressing until an action happens. Looks like indicator is not needed. But please let me know what you think about a visual indicator after testing on dev.

**Comment by david-a-wheeler** [comment]
FYI, this might help. I did a little searching on how to disable the default text-selection behavior when the text is showing in non-edit mode, and found this discussion on how to set the CSS:

https://stackoverflow.com/questions/66151807/how-to-disable-text-selection-on-touch-press

I think normal copy/paste should work once a field is opened in "simple edit" mode. However, when it's in "pretty display" mode (and might get alt+click including long-click), it should (1) switch to edit mode on normal click or (2) switch to & use the text selector on alt+click (including long-click).

**Comment by expln** [comment]
> Sorry about that, I'd hoped it'd be easy. But I think the ease of explaining "long click == alt-click" is worth it. Making it all "as similar as possible" makes things easier to explain & switch between.

No worries! You are right, it appeared to be a very useful feature - I like it and now I use it instead of pressing Alt key :)

**Comment by david-a-wheeler** [comment]
I think this is ready-on-dev!

A few notes:

* The H/P/G selector works great. Sure, it'd be nice if it responded to a bigger area, but that's minor issue; pinch in and out resolve it fine.
* If you do this on statements, the smartphone selectors fight over text selection, but it doesn't *prevent* its use. Dealing with that might be useful later refinement, but it's already much better than v10.
* I still think that long-click to edit description shouldn't happen - it should only require a simple click to edit just like every other simple edit. But at least it *is* editable on a smartphone, and that was my point.

In short, this is *great*. Don't let my comments on nits divert you from the bigger picture, this is a *great* addition.

**Comment by expln** [comment]
The default text-selection behavior is disabled on touchscreens as per the link you've provided.
